### PR TITLE
core/validatorapi: validatorapi implementation builder registration

### PR DESCRIPTION
**Pull request by ciaranmcveigh5**
**WORK IN PROGRESS**

- Updates to router
- Updates to vapi

category: feature
ticket: #849

### CODE REVIEW

File: eth2util/signing/signing.go
Comment (ciaranmcveigh5) [inline eth2util/signing/signing.go:36]: This requires an additional update in go-eth2-client

PR has been raised waiting for Jim to review

https://github.com/attestantio/go-eth2-client/pull/23

File: core/validatorapi/router.go
Comment (ciaranmcveigh5) [inline core/validatorapi/router.go:133]: The implementation in go-eth2-client is submit registrations as an array for all validators on a given node

https://github.com/attestantio/go-eth2-client/blob/c05328b73d87b8bcca8cab74f97c471fd85e5f3a/http/submitvalidatorregistrations.go

File: core/validatorapi/validatorapi.go
Comment (ciaranmcveigh5) [inline core/validatorapi/validatorapi.go:502]: Not used - dont think this is needed?

File: core/validatorapi/validatorapi.go
Comment (ciaranmcveigh5) [inline core/validatorapi/validatorapi.go:538]: Normally would used the Root() function 

This is implemented in go-eth2-client yet - PR has been raised 

Under the hood Root() just calles V1.Message.HashTreeRoot() so can use this in the interim until the PR is merged by Jim

https://github.com/attestantio/go-eth2-client/pull/23

File: core/validatorapi/validatorapi.go
Comment (ciaranmcveigh5) [inline core/validatorapi/validatorapi.go:550]: Need to change this to either from the timestamp or get current epoch and use the first slot of the next epoch

using hardcoded currently just to get things working

File: core/validatorapi/validatorapi.go
Comment (corverroos) [inline core/validatorapi/validatorapi.go:554]: So this is actually a problem. The pubkey included in the message will be the "pubshare" not the "group/root pubkey" since the VC only knows about it's pubshare not about the group pubkey. 

That means the signatures cannot be aggregated since the pubshare of each VC is different.... 😢 

Add a TODO maybe "this is the pubshare not the group pubkey, oh oh"

Comment (corverroos) [reply core/validatorapi/validatorapi.go:554]: and btw,`verifyParSig` requires the group pubkey, not the pubshare.

File: core/validatorapi/validatorapi.go
Comment (corverroos) [reply core/validatorapi/validatorapi.go:502]: can remove yeah

File: go.sum
Comment (corverroos) [inline go.sum:64]: not sure why this was added?

**Review by corverroos — APPROVED** [review]

**Pull request merged**
